llvm/lib/Target/AIE/AIELegalizerInfo.cpp

```diff
@@ -493,6 +493,24 @@ AIELegalizerInfo::AIELegalizerInfo(const AIEBaseSubtarget &ST) {
         .clampMaxNumElements(0, S16, 32)
         .clampMaxNumElements(0, S32, 16)
         .custom();
+
+    getActionDefinitionsBuilder(G_SHUFFLE_VECTOR)
+        .unsupportedIf(IsNotValidDestinationVector)
```

Comment:
There is a final `.lower()`, so no need for the `lowerIf` anymore.
Do we need the clampMinNumElements?

Reply:
The `clampMinNumElements` do something useful in that it avoids generating bad IR. Lowering it directly will cause them to crash since you are moving the buck down. It is not that useful, but actually what it should do is clamp the arguments instead. I will change it to that.

`lowerIf` is mixed. I could remove most of them, but I also see the point in keeping them there as documentation. What a shufflevector looks like isn't very clear and it is non-obvious to someone new to LLVM that it can turn scalars into vectors or that there is a "canonical" vector with different properties. Lowering them separately makes it clear that they are different cases. So I am a bit mixed on that.

Reply:
I removed all but the `.lower()` action and all regression tests pass, except [xfail-legalize-shufflevector.mir](https://github.com/Xilinx/llvm-aie/blob/aie-public/llvm/test/CodeGen/AIE/GlobalISel/xfail-legalize-shufflevector.mir) . Seems like there are missing tests for the `clampMinNumElements` cases?
Could you either add a test that shows these clamps are needed, or remove all other legalization actions.

The xfail test fails because `.lower` actually generates the expected MIR instead of crashing. We can remove the RUN line for aie2 for that test

Reply:
Yeah, fair enough. Clamping for the input vector can be useful, which is different than what it does now, however recently 64-bit vectors support was added in #77 which is the main missing part. I'll remove it for now, but keep in mind that this could lead to errors in the form of: `cannot legalize: %0:_<1 x i16> = G_SOME_FUNC`.

I am keeping two checks: canonical and the check whether it has an AIE2 vector output. The former does weird things to the Combiner that I would like people to be aware and the latter gives some flexibility for future architectures.